### DISCUSSION TITLE
How to deal with memory leaks using sp.IdlFile

### DISCUSSION THREAD

**Discussion by rebeccaringuette** [spacepy/spacepy#690] [Support and Q&A]
The SpacePy documentation does not describe how to close a file opened by the command...

### Minimal example to reproduce issue:
```py
import spacepy.pybats as sp
mhd = sp.IdlFile(f)
mhd.close()
```
where f is the complete path and filename of a SWMF GM .out file.

### Error message/Traceback:
```py
AttributeError: 'IdlFile' object has no attribute 'close'
```

### OS, Python version, and dependency version information:
```py
Linux-5.4.238-148.346.amzn2.x86_64-x86_64-with-glibc2.35
sys.version_info(major=3, minor=10, micro=10, releaselevel='final', serial=0)
numpy=1.23.5
scipy=1.10.1
matplotlib=3.7.1
```

### Version of SpacePy
Current version on GitHub via git clone. (May 2023)

**Comment by jtniehof** [comment]
@spacecataz is the definitive source here, but it looks like the various methods of IdlFile open the file, perform whatever reads are necessary, and then close--so the file isn't being left open. That does mean the data read are left in memory in the IdlFile object, so making a bunch of IdlFile objects, reading data in, and not removing those objects will lead to memory exhaustion.

**Comment by rebeccaringuette** [comment]
That is exactly the problem, actually. I added some garbage collection after those commands and the memory no longer crashes.
```py
import gc
mhd = 0
gc.collect()
```
...but this is so slow. Is there a faster way to just get the datetime object from the file? The code snippet below simply retrieves the datetime object and converts it to seconds since midnight, but it takes several seconds per file, and I have a lot of files.
```py
        @vectorize
        def get_datetime(filename, filedate):
            '''
            Parameters
            ----------
            filename : string
                Name of .out file with complete path
            filedate : datetime object
                Datetime object corresponding to midnight of the first day.

            Returns
            -------
            Number of seconds since filedate for data in given file.

            '''
            file_dt = sp.IdlFile(filename, sort_unstructured=False).attrs[
                'time'].replace(tzinfo=timezone.utc)
            gc.collect()  # remove file contents from memory?
            print(filename, file_dt)
            return (file_dt - filedate).total_seconds()/3600.
```

**Comment by spacecataz** [comment]
Lots to respond to here. I'm going to start by converting this to a conversation, as the original topic is now not relevant.

**Reply by rebeccaringuette** [reply]
I renamed the topic, too.

**Reply by spacecataz** [reply]
I'm frantically typing a reply!

**Comment by spacecataz** [comment]
Okay, taking things in-turn:

- `.close` syntax is only relevant if the returned object was some sort of data stream (e.g., a true python file object). But they're not, so no method is required. Behind the scenes, here is what's happening: the `.out` file is opened, the contents read, and then the file closed. In other words, you're bringing everything to memory. The only exception is for `.outs` files (note the `s`!), which is many `.out` files concatenated together (this is a standard operation for SWMF output via the PostProc.pl script.) In the case of `.outs` files, only the first frame is read. You then switch frames explicitly to avoid situations where you're eating up memory.

- The garbage collection issue you're having is curious to me. My typical work flow is to cycle over many many gigabytes of files, doing analysis in-turn. So long as I'm reusing the variable name on each iteration, I do not run into memory issues. Perhaps there is something happening with your `@vectorize` decorator...

- Now, the core issue: getting times for a bunch of files! Indeed, opening the full file just to get a time stamp is very inefficient. Allow me to propose some alternatives.

1) Use `parse_filename_time`:
 ```
>>> from spacepy import pybats
>>>  file1 = 'z=0_mhd_2_e20121008-010000-000_20121009-120000-000.outs'
(None,
 None,
 [datetime.datetime(2012, 10, 8, 1, 0), datetime.datetime(2012, 10, 9, 12, 0)])
```
Check the docstring on that guy.

2) If your files are combined as a `.outs` file, the first read will load only the first time frame, but the attributes will have *all* of the datetimes as an array. This is as fast as opening 1 file, but scans the headers of each contained file without reading in all the data.
```
>>> mhd1 = pybats.IdlFile(file1)
>>> mhd1.attrs['times']
array([datetime.datetime(2012, 10, 8, 1, 0),
       datetime.datetime(2012, 10, 8, 1, 1),
       datetime.datetime(2012, 10, 8, 1, 2), ...,
       datetime.datetime(2012, 10, 9, 11, 58),
       datetime.datetime(2012, 10, 9, 11, 59),
       datetime.datetime(2012, 10, 9, 12, 0)], dtype=object)
```

3) Finally, there are some internal use functions (ones that start with underscores) that would allow you to scan file headers quickly and get attribute information without loading data. I don't necessarily recommend this, but perhaps a convenience function that is visible to users would be a good thing to include in the next update...

Let me know if all this helps.

**Reply by rebeccaringuette** [reply]
Thanks, Jon. Where is the parse_filename_time function? Just pybats? I don't see it in the documentation.

**Reply by spacecataz** [reply]
I'm a dan, not a jon.

It's in pybats.
Docs in the docstring:
```
from spacepy import pybats
pybats.parse_filename_time?  # from within ipython
```

**Comment by rebeccaringuette** [comment]
Thanks, Dan. That made a HUGE difference in timing! I still need to test how this issue affects interpolation through the data, so I'll comment here later on that. Code that worked (from different parts of the script, so indentation looks weird here):
```py
import spacepy.pybats as sp
from numpy import vectorize

        @vectorize
        def get_datetime(filename, filedate):
            '''
            Parameters
            ----------
            filename : string
                Name of .out file with complete path
            filedate : datetime object
                Datetime object corresponding to midnight of the first day.

            Returns
            -------
            Number of seconds since filedate for data in given file.

            '''
            file_dt = sp.parse_filename_time(filename)[-1].replace(
                tzinfo=timezone.utc)
            return (file_dt - filedate).total_seconds()/3600.
